### ISSUE TITLE
Problems With Solhint Plugins

### ISSUE THREAD

**Issue by Jaime-Iglesias** [juanfranblanco/vscode-solidity#169]
I'm trying to use the rules on a custom solhint plugin by changing the `.solhint.json` configuration file.
The rules themselves work as I can lint the file manually using the terminal and I get the expected errors, but I cannot manage to make the Solidity plugin itself detect these custom rules.

I have changed the settings to explicitly use `solhint` as per the documentation i.e

```
"solidity.linter": "solium",
"solidity.soliumRules": {
    "quotes": ["error", "double"],
    "indentation": ["error", 4]
},
```

And I have changed the `.solhint.json` as per the Solhint docs too.

Any idea of what I am doing wrong?

**Comment by juanfranblanco** [comment]
@Jaime-Iglesias is the .solhint.json file at root?

**Comment by Jaime-Iglesias** [comment]
@juanfranblanco yep

![structure](https://user-images.githubusercontent.com/22358726/78137385-21503680-741d-11ea-8061-6337e2adeb14.png)

**Comment by juanfranblanco** [comment]
@Jaime-Iglesias  First test without .solhint.json: 
![image](https://user-images.githubusercontent.com/562371/78138115-5446fa00-741e-11ea-976b-0641b85502d6.png)
![image](https://user-images.githubusercontent.com/562371/78138168-69bc2400-741e-11ea-9b7c-1cef96b228e5.png)

**Comment by Jaime-Iglesias** [comment]
@juanfranblanco without the `.solhint.json` and with it, the rules in the `settings` work just fine

![test](https://user-images.githubusercontent.com/22358726/78138536-f6ff7880-741e-11ea-8d58-9827426dbf03.png)

![test](https://user-images.githubusercontent.com/22358726/78138622-21e9cc80-741f-11ea-93a7-d54738d0daa9.png)

**Comment by juanfranblanco** [comment]
Ok I have tested it again with a file:

![image](https://user-images.githubusercontent.com/562371/78138823-74c38400-741f-11ea-9cac-a9f0958d434d.png)

![image](https://user-images.githubusercontent.com/562371/78138872-8573fa00-741f-11ea-971b-b4cf1104585c.png)

![image](https://user-images.githubusercontent.com/562371/78138930-96bd0680-741f-11ea-9c03-5734459ae6d2.png)

**Comment by juanfranblanco** [comment]
And to make sure it just works with the file:

![image](https://user-images.githubusercontent.com/562371/78139363-4c885500-7420-11ea-81df-e980b4de14a5.png)
![image](https://user-images.githubusercontent.com/562371/78139408-6033bb80-7420-11ea-90bf-a3236ab123d0.png)
![image](https://user-images.githubusercontent.com/562371/78139449-6fb30480-7420-11ea-991b-4b760cb890fa.png)


What it may have happened is that you may need to restart vscode due to a problem before. Try again by adding the file.

**Comment by Jaime-Iglesias** [comment]
hmmm, the problem does not seem to be with the file but with the extension not being able to detech the solhint plugin since it's clear that the rules in the `.solhint.json` that do not belong to the custom plugin are being enforced

![test](https://user-images.githubusercontent.com/22358726/78139917-20210880-7421-11ea-8de1-7dcdb6e9772d.png)

![test](https://user-images.githubusercontent.com/22358726/78139841-02ec3a00-7421-11ea-9cdb-abd2279a04f9.png)

![test](https://user-images.githubusercontent.com/22358726/78139751-e6500200-7420-11ea-9cc5-927fe97501e1.png)

**Comment by juanfranblanco** [comment]
Ahh I got you, yes the extension does not recognize plugins, where do you host your plugins?

**Comment by Jaime-Iglesias** [comment]
> Ahh I got you, yes the extension does not recognize plugins, where do you host your plugins?

What do you mean where I host ? I install them locally after publishing them (as per the solhint docs), hope that answers the question haha

**Comment by juanfranblanco** [comment]
Yes if you install them locally the extension might not be able to find them. I do need to do some research on this, or raise it with the solium guys.

**Comment by Jaime-Iglesias** [comment]
> Yes if you install them locally the extension might not be able to find them. I do need to do some research on this, or raise it with the solium guys.

So if they are globally installed they do work ? (haven't tried that option yet) also thank you

**Comment by juanfranblanco** [comment]
At the moment the extension does not load the plugins section, i can add that and check it globally.

**Comment by Jaime-Iglesias** [comment]
> At the moment the extension does not load the plugins section, i can add that and check it globally.

That would be great!

Ofc, ideally, loading them locally would be huge but progress is progress, thx again.

**Comment by juanfranblanco** [comment]
@Jaime-Iglesias quick test, it does not work i have added the no-foos rules install your package globally and nothing, it will have to be find and loaded dynamically by the extension into solhint, similar to prettier.

**Comment by Jaime-Iglesias** [comment]
> @Jaime-Iglesias quick test, it does not work i have added the no-foos rules install your package globally and nothing, it will have to be find and loaded dynamically by the extension into solhint, similar to prettier.

@juanfranblanco I just tried it myself too but no luck either, it's weird since `solhint` supports plugins out of the box, I guess it's cause by how the Extension itself manages all these packages

**Comment by juanfranblanco** [comment]
@Jaime-Iglesias looking at the source would have been faster https://github.com/protofire/solhint/blob/32ec781509c9b76db325d0c2db91c8c478441d6a/lib/rules/index.js#L53

there is a require there, which won't be triggered as the extensions cannot access those.

**Comment by Jaime-Iglesias** [comment]
> @Jaime-Iglesias looking at the source would have been faster https://github.com/protofire/solhint/blob/32ec781509c9b76db325d0c2db91c8c478441d6a/lib/rules/index.js#L53
> 
> there is a require there, which won't be triggered as the extensions cannot access those. Prettier provides a path setting that allows for the discovery of plugins, something they could implement.

I see, so it's a `solhint` problem after all - I wonder if they will manage to fix it, I don't know active the project is.

**Comment by juanfranblanco** [comment]
@Jaime-Iglesias  I have added an issue for this https://github.com/protofire/solhint/issues/206

**Comment by Jaime-Iglesias** [comment]
> @Jaime-Iglesias I have added an issue for this [protofire/solhint#206](https://github.com/protofire/solhint/issues/206)

@juanfranblanco much appreciated, hopefully they will be able to support this, as I think it defenitely can bring a lot of value for both projects.

**Comment by Jaime-Iglesias** [comment]
Hey @juanfranblanco I hope you are doing well,

Is there any update regarding this issue and https://github.com/juanfranblanco/vscode-solidity/issues/170 ?

stay safe !

**Comment by Jaime-Iglesias** [comment]
friendly ping @juanfranblanco

**Comment by juanfranblanco** [comment]
@Jaime-Iglesias thanks for the ping!! I do need these sometimes as notifications get lost (you can find me in the Nethereum gitter). The issue is still open with the solhint guys / gals, the best way is to chase that issue there, once is fixed on their side it could be integrated here.

**Comment by Jaime-Iglesias** [comment]
> @Jaime-Iglesias thanks for the ping!! I do need this as sometimes notifications get lost (you can find me in the Nethereum gitter if you need anything). The issue is still open with the solhint guys / gals, the best way is to chase that issue there, once is fixed on their side it could be integrated here.

Ah I see, I thought it was depending on you - sorry about that hahaha! Thanks and stay safe !
